### ISSUE TITLE
SAML client and authorization grants

### ISSUE THREAD

**Issue by F21** [bshaffer/oauth2-server-php#96]
Was just reading about SAML and found that they can also be used to authenticate and authorize OAuth 2 tokens.

The flow seems to be quite similar to how JWT works, so we should be quite familiar with it. 

IETF Draft: http://tools.ietf.org/html/draft-ietf-oauth-saml2-bearer-16

This is not a must have or anything urgent, but would be nice to implement when any of us have some extra time. The flow should be very similar if not identical to the flow for JWT tokens, so it shouldn't be too complicated.

Interestingly, it also has a client authentication mode similar to JWT's client authentication in #27, so perhaps we can look at a general way to abstract the `AuthorizationCode` grant type to support JWTs and SAML.

**Comment by aacotroneo** [comment]
Hi, Any updates on this one? I'm interested. I may take a look.

**Comment by F21** [comment]
@aacotroneo Unfortunately, no. Would love to see a PR if you're interested! :smile:

**Comment by aacotroneo** [comment]
@F21 Of course! :) I haven't even started, but It shouldn't take long (I hope). I'll start by reading the draft!

**Comment by bshaffer** [comment]
@aacotroneo that would be amazing

**Comment by aacotroneo** [comment]
Well, I'm working on it! 

Much of the validation the oauth2-server needs to do is pretty similar to what a saml service provider does. So, I thought about making an adapter to use onelogin/php-saml which handles all the hard xml validation stuff and has lots of unit tests.

I'm dealing with saml configuration right now. I'll try to make a simple case work, and then see how we can refactor it.

**Comment by bshaffer** [comment]
Even if you have a demo app and enough material to add a cookbook article to the docs, that would also be immensely helpful

**Comment by aacotroneo** [comment]
#510

**Comment by aacotroneo** [comment]
@bshaffer Here is a sample implementation of your server using Slim framework (i saw a request for that). You'll see that it also implements the saml2-bearer grant type, but there's not much to see there. It just adds the grant type with a proper configuration array. I'll see if I find a way to make a test client without needing to setup a real IDP.
